File: internal/providers/providers.go
// Package providers shells out to whichever local LLM CLI the user
// has configured. The analyzer treats this as an opaque
// string-in / string-out boundary.
//
// Supported providers (auto-detect order):
//
//	claude   — Anthropic Claude Code CLI
//	gemini   — Google Gemini CLI
//	openai   — OpenAI CLI
//	ollama   — local Ollama
//	generic  — any CLI specified via WATCHDOG_LLM_CMD
//
// The child process always receives WATCHDOG_DISABLE=1 in its env so
// any nested hook the LLM might trigger short-circuits and does not
// recursively re-invoke this analyzer.
package providers

import (
	"bytes"
	"context"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
	"time"

	"github.com/Maxlemore97/watchdog/internal/parsers"
)

var ValidProviders = map[string]bool{
	"claude": true, "gemini": true, "openai": true, "ollama": true, "generic": true,
}

var AutoDetectOrder = []string{"claude", "gemini", "openai", "ollama"}

const DefaultTimeout = 60 * time.Second

// Provider describes one configured LLM CLI.
type Provider struct {
	Name         string
	Bin          string
	DefaultModel string
	Invoke       func(prompt string, cfg Config) (string, error)
}

// Config is the per-invocation settings derived from the env.
type Config struct {
	Bin          string
	Model        string
	SystemPrompt string
	AppendSystem bool
	Timeout      time.Duration
	Cmd          string // only for generic
}

// Registry holds all providers, addressable by name.
var Registry = map[string]Provider{
	"claude": {
		Name: "claude", Bin: "claude", DefaultModel: "claude-haiku-4-5-20251001",
		Invoke: invokeClaude,
	},
	"gemini": {
		Name: "gemini", Bin: "gemini", DefaultModel: "gemini-2.5-flash",
		Invoke: invokeGemini,
	},
	"openai": {
		Name: "openai", Bin: "openai", DefaultModel: "gpt-4.1-mini",
		Invoke: invokeOpenAI,
	},
	"ollama": {
		Name: "ollama", Bin: "ollama", DefaultModel: "llama3.1",
		Invoke: invokeOllama,
	},
	"generic": {
		Name: "generic", Bin: "", DefaultModel: "generic",
		Invoke: invokeGeneric,
	},
}

// ErrNoProvider is returned when no usable LLM CLI is available.
var ErrNoProvider = errors.New("providers: no usable LLM CLI on PATH")

// ResolveProvider picks a provider per WATCHDOG_LLM_PROVIDER, falling
// back to PATH auto-detect. Returns ErrNoProvider when none usable.
func ResolveProvider() (Provider, error) {
	raw := strings.ToLower(strings.TrimSpace(os.Getenv("WATCHDOG_LLM_PROVIDER")))
	if raw == "" || raw == "auto" {
		return autoDetect()
	}
	if !ValidProviders[raw] {
		return autoDetect()
	}
	prov := Registry[raw]
	if prov.Name == "generic" {
		// generic always returns; its invoke rejects when
		// WATCHDOG_LLM_CMD is unset/unrunnable.
		return prov, nil
	}
	if _, err := exec.LookPath(prov.Bin); err != nil {
		return Provider{}, ErrNoProvider
	}
	return prov, nil
}

func autoDetect() (Provider, error) {
	for _, name := range AutoDetectOrder {
		prov := Registry[name]
		if _, err := exec.LookPath(prov.Bin); err == nil {
			return prov, nil
		}
	}
	return Provider{}, ErrNoProvider
}

// BuildConfig assembles per-call settings from env.
func BuildConfig(prov Provider, systemPrompt string) Config {
	bin := prov.Bin
	if override := strings.TrimSpace(os.Getenv("WATCHDOG_LLM_BIN")); override != "" {
		bin = override
	}
	model := strings.TrimSpace(os.Getenv("WATCHDOG_LLM_MODEL"))
	if model == "" {
		model = prov.DefaultModel
	}
	timeout := DefaultTimeout
	if raw := strings.TrimSpace(os.Getenv("WATCHDOG_LLM_TIMEOUT")); raw != "" {
		if v, err := strconv.ParseFloat(raw, 64); err == nil {
			timeout = time.Duration(v * float64(time.Second))
		}
	}
	appendSystem := true
	switch strings.ToLower(strings.TrimSpace(os.Getenv("WATCHDOG_LLM_APPEND_SYSTEM"))) {
	case "0", "false", "no", "off":
		appendSystem = false
	}
	return Config{
		Bin:          bin,
		Model:        model,
		SystemPrompt: systemPrompt,
		AppendSystem: appendSystem,
		Timeout:      timeout,
		Cmd:          os.Getenv("WATCHDOG_LLM_CMD"),
	}
}

// InvokeLLM resolves the provider, builds the config, and runs the
// CLI. Returns (stdout, provider, config, error). On no-provider:
// returns ErrNoProvider with zero-valued provider/config.
func InvokeLLM(prompt, systemPrompt string) (string, Provider, Config, error) {
	prov, err := ResolveProvider()
	if err != nil {
		return "", Provider{}, Config{}, err
	}
	cfg := BuildConfig(prov, systemPrompt)
	output, err := prov.Invoke(prompt, cfg)
	return output, prov, cfg, err
}

// ---------- per-provider invocations ------------------------------

func childEnv() []string {
	env := os.Environ()
	out := make([]string, 0, len(env)+1)
	for _, kv := range env {
		if !strings.HasPrefix(kv, "WATCHDOG_DISABLE=") {
			out = append(out, kv)
		}
	}
	out = append(out, "WATCHDOG_DISABLE=1")
	return out
}

// boundedBuffer caps stderr capture from LLM CLIs so a misbehaving
// provider (or a non-LLM binary the user pointed WATCHDOG_LLM_CMD at)
// can't OOM watchdog by flooding stderr.
type boundedBuffer struct {
	limit int
	buf   bytes.Buffer
}

func (b *boundedBuffer) Write(p []byte) (int, error) {
	remain := b.limit - b.buf.Len()
	if remain <= 0 {
		return len(p), nil
	}
	if len(p) > remain {
		b.buf.Write(p[:remain])
		return len(p), nil
	}
	return b.buf.Write(p)
}

func (b *boundedBuffer) String() string { return b.buf.String() }

const stderrCapBytes = 64 * 1024

func runCmd(ctx context.Context, name string, args []string, stdin string) (string, error) {
	cmd := exec.CommandContext(ctx, name, args...)
	cmd.Stdin = strings.NewReader(stdin)
	cmd.Env = childEnv()
	var stdout bytes.Buffer
	stderr := &boundedBuffer{limit: stderrCapBytes}
	cmd.Stdout = &stdout
	cmd.Stderr = stderr
	if err := cmd.Run(); err != nil {
		return "", fmt.Errorf("%s: %w (stderr: %s)", name, err, truncate(stderr.String(), 200))
	}
	return stdout.String(), nil
}

func invokeClaude(prompt string, cfg Config) (string, error) {
	if _, err := exec.LookPath(cfg.Bin); err != nil {
		return "", ErrNoProvider
	}
	args := []string{
		"-p",
		"--model", cfg.Model,
		"--output-format", "json",
		"--max-turns", "1",
		"--allowed-tools", "",
	}
	if cfg.AppendSystem {
		args = append(args, "--append-system-prompt", cfg.SystemPrompt)
	}
	ctx, cancel := context.WithTimeout(context.Background(), cfg.Timeout)
	defer cancel()
	return runCmd(ctx, cfg.Bin, args, prompt)
}

func invokeGemini(prompt string, cfg Config) (string, error) {
	if _, err := exec.LookPath(cfg.Bin); err != nil {
		return "", ErrNoProvider
	}
	body := combineSystemUser(prompt, cfg)
	ctx, cancel := context.WithTimeout(context.Background(), cfg.Timeout)
	defer cancel()
	return runCmd(ctx, cfg.Bin, []string{"-m", cfg.Model}, body)
}

func invokeOpenAI(prompt string, cfg Config) (string, error) {
	if _, err := exec.LookPath(cfg.Bin); err != nil {
		return "", ErrNoProvider
	}
	args := []string{"api", "chat.completions.create", "-m", cfg.Model}
	if cfg.AppendSystem {
		args = append(args, "-g", "system", cfg.SystemPrompt)
	}
	args = append(args, "-g", "user", prompt)
	ctx, cancel := context.WithTimeout(context.Background(), cfg.Timeout)
	defer cancel()
	return runCmd(ctx, cfg.Bin, args, "")
}

func invokeOllama(prompt string, cfg Config) (string, error) {
	if _, err := exec.LookPath(cfg.Bin); err != nil {
		return "", ErrNoProvider
	}
	body := combineSystemUser(prompt, cfg)
	ctx, cancel := context.WithTimeout(context.Background(), cfg.Timeout)
	defer cancel()
	return runCmd(ctx, cfg.Bin, []string{"run", cfg.Model}, body)
}

func invokeGeneric(prompt string, cfg Config) (string, error) {
	if cfg.Cmd == "" {
		return "", ErrNoProvider
	}
	argv, err := parsers.Tokenize(cfg.Cmd)
	if err != nil || len(argv) == 0 {
		return "", ErrNoProvider
	}
	if _, err := exec.LookPath(argv[0]); err != nil {
		return "", ErrNoProvider
	}
	body := combineSystemUser(prompt, cfg)
	ctx, cancel := context.WithTimeout(context.Background(), cfg.Timeout)
	defer cancel()
	return runCmd(ctx, argv[0], argv[1:], body)
}

func combineSystemUser(prompt string, cfg Config) string {
	if !cfg.AppendSystem || cfg.SystemPrompt == "" {
		return prompt
	}
	return "=== SYSTEM ===\n" + cfg.SystemPrompt + "\n=== USER ===\n" + prompt + "\n"
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n]
}

File: watchdog_core/providers.py
"""LLM provider registry for the source-review analyzer.

Watchdog needs *some* local LLM CLI to perform source analysis on
fetched package/plugin files. This module abstracts that CLI behind a
small Provider protocol so users can plug in whatever model they have
available locally:

- `claude`   — Anthropic Claude Code CLI (`claude -p`)
- `gemini`   — Google Gemini CLI (`gemini`)
- `openai`   — OpenAI CLI (`openai`)
- `ollama`   — local Ollama (`ollama run <model>`)
- `generic`  — any CLI specified via WATCHDOG_LLM_CMD; prompt piped via stdin

Selection precedence:
1. `WATCHDOG_LLM_PROVIDER` env var (claude/gemini/openai/ollama/generic).
2. Auto-detect: first of (claude, gemini, openai, ollama) found on PATH.

`generic` is never auto-selected; it must be explicitly requested.

All providers set `WATCHDOG_DISABLE=1` in the child env so any nested
agent session's hooks short-circuit and do not recursively re-invoke
the analyzer.
"""
from __future__ import annotations

import os
import shlex
import shutil
import subprocess
from dataclasses import dataclass
from typing import Callable, Optional

VALID_PROVIDERS = {"claude", "gemini", "openai", "ollama", "generic"}
AUTO_DETECT_ORDER = ("claude", "gemini", "openai", "ollama")

DEFAULT_TIMEOUT = 60.0


@dataclass(frozen=True)
class Provider:
    name: str
    bin: str
    default_model: str
    invoke: Callable[[str, "ProviderConfig"], Optional[str]]


@dataclass(frozen=True)
class ProviderConfig:
    bin: str
    model: str
    system_prompt: str
    append_system: bool
    timeout: float
    cmd: str  # only used by `generic`


def _child_env() -> dict[str, str]:
    env = os.environ.copy()
    env["WATCHDOG_DISABLE"] = "1"
    return env


def _run(cmd: list[str], stdin: str, timeout: float) -> Optional[str]:
    try:
        proc = subprocess.run(
            cmd,
            input=stdin,
            capture_output=True,
            text=True,
            timeout=timeout,
            env=_child_env(),
        )
    except (subprocess.TimeoutExpired, FileNotFoundError, OSError):
        return None
    if proc.returncode != 0:
        return None
    return proc.stdout


# ---------- claude --------------------------------------------------------

def _invoke_claude(prompt: str, cfg: ProviderConfig) -> Optional[str]:
    if not shutil.which(cfg.bin):
        return None
    cmd = [
        cfg.bin,
        "-p",
        "--model",
        cfg.model,
        "--output-format",
        "json",
        "--max-turns",
        "1",
        "--allowed-tools",
        "",
    ]
    if cfg.append_system:
        cmd += ["--append-system-prompt", cfg.system_prompt]
    return _run(cmd, stdin=prompt, timeout=cfg.timeout)


# ---------- gemini --------------------------------------------------------

def _invoke_gemini(prompt: str, cfg: ProviderConfig) -> Optional[str]:
    """Google Gemini CLI. Pipes a combined system+user prompt via stdin
    because the public `gemini` CLI's flag shape has shifted across
    versions; stdin is the stable surface."""
    if not shutil.which(cfg.bin):
        return None
    body = _combine_system_user(prompt, cfg)
    cmd = [cfg.bin, "-m", cfg.model]
    return _run(cmd, stdin=body, timeout=cfg.timeout)


# ---------- openai --------------------------------------------------------

def _invoke_openai(prompt: str, cfg: ProviderConfig) -> Optional[str]:
    """OpenAI CLI. Uses the `openai api chat.completions.create` form
    with system+user messages. Newer CLI variants accept stdin; we use
    explicit message args for deterministic shape."""
    if not shutil.which(cfg.bin):
        return None
    cmd = [
        cfg.bin,
        "api",
        "chat.completions.create",
        "-m",
        cfg.model,
    ]
    if cfg.append_system:
        cmd += ["-g", "system", cfg.system_prompt]
    cmd += ["-g", "user", prompt]
    return _run(cmd, stdin="", timeout=cfg.timeout)


# ---------- ollama --------------------------------------------------------

def _invoke_ollama(prompt: str, cfg: ProviderConfig) -> Optional[str]:
    """Local Ollama. `ollama run <model>` reads user input via stdin
    and emits the model response on stdout."""
    if not shutil.which(cfg.bin):
        return None
    body = _combine_system_user(prompt, cfg)
    cmd = [cfg.bin, "run", cfg.model]
    return _run(cmd, stdin=body, timeout=cfg.timeout)


# ---------- generic -------------------------------------------------------

def _invoke_generic(prompt: str, cfg: ProviderConfig) -> Optional[str]:
    """Run any user-specified CLI via WATCHDOG_LLM_CMD. The command is
    shlex-split (no shell interpolation). System+user prompt is piped
    via stdin; raw stdout is returned for verdict extraction."""
    if not cfg.cmd:
        return None
    try:
        argv = shlex.split(cfg.cmd)
    except ValueError:
        return None
    if not argv or not shutil.which(argv[0]):
        return None
    body = _combine_system_user(prompt, cfg)
    return _run(argv, stdin=body, timeout=cfg.timeout)


# ---------- helpers -------------------------------------------------------

def _combine_system_user(prompt: str, cfg: ProviderConfig) -> str:
    """Prepend the system prompt to the user prompt when the provider
    cannot pass it via a dedicated flag. Section markers help models
    that have no native system-role handling."""
    if not cfg.append_system or not cfg.system_prompt:
        return prompt
    return (
        "=== SYSTEM ===\n"
        f"{cfg.system_prompt}\n"
        "=== USER ===\n"
        f"{prompt}\n"
    )


# ---------- registry & resolution -----------------------------------------

REGISTRY: dict[str, Provider] = {
    "claude": Provider("claude", "claude", "claude-haiku-4-5-20251001", _invoke_claude),
    "gemini": Provider("gemini", "gemini", "gemini-2.5-flash", _invoke_gemini),
    "openai": Provider("openai", "openai", "gpt-4.1-mini", _invoke_openai),
    "ollama": Provider("ollama", "ollama", "llama3.1", _invoke_ollama),
    "generic": Provider("generic", "", "generic", _invoke_generic),
}


def auto_detect() -> Optional[Provider]:
    for name in AUTO_DETECT_ORDER:
        prov = REGISTRY[name]
        if shutil.which(prov.bin):
            return prov
    return None


def resolve_provider() -> Optional[Provider]:
    """Pick a provider per WATCHDOG_LLM_PROVIDER or auto-detect.

    Returns None when no usable provider is available (e.g. provider
    pinned but its CLI missing, or `auto` and nothing on PATH). Caller
    treats None as "LLM step unavailable" and falls back to its
    offline_decision policy."""
    raw = os.environ.get("WATCHDOG_LLM_PROVIDER", "").strip().lower()
    if not raw or raw == "auto":
        return auto_detect()
    if raw not in VALID_PROVIDERS:
        return auto_detect()
    prov = REGISTRY[raw]
    if prov.name == "generic":
        # generic always returns from resolve_provider; the invoke
        # function rejects if WATCHDOG_LLM_CMD is unset/unrunnable.
        return prov
    if not shutil.which(prov.bin):
        return None
    return prov


def build_config(provider: Provider, system_prompt: str) -> ProviderConfig:
    bin_override = os.environ.get("WATCHDOG_LLM_BIN", "").strip()
    model = os.environ.get("WATCHDOG_LLM_MODEL", "").strip() or provider.default_model
    timeout_raw = os.environ.get("WATCHDOG_LLM_TIMEOUT", "").strip()
    try:
        timeout = float(timeout_raw) if timeout_raw else DEFAULT_TIMEOUT
    except ValueError:
        timeout = DEFAULT_TIMEOUT
    append_system = os.environ.get("WATCHDOG_LLM_APPEND_SYSTEM", "1").strip().lower() not in {
        "0",
        "false",
        "no",
        "off",
    }
    cmd = os.environ.get("WATCHDOG_LLM_CMD", "")
    return ProviderConfig(
        bin=bin_override or provider.bin,
        model=model,
        system_prompt=system_prompt,
        append_system=append_system,
        timeout=timeout,
        cmd=cmd,
    )


def invoke_llm(prompt: str, system_prompt: str) -> tuple[Optional[str], Optional[Provider], Optional[ProviderConfig]]:
    """Resolve provider, build config, invoke. Returns (stdout, provider, config).

    Returns (None, None, None) when no provider is available.
    Provider and config are returned even on invocation failure so the
    caller can include them in cache keys and diagnostics."""
    provider = resolve_provider()
    if provider is None:
        return None, None, None
    cfg = build_config(provider, system_prompt)
    output = provider.invoke(prompt, cfg)
    return output, provider, cfg
